index/scorch/snapshot_rollback.go

```diff
@@ -105,69 +116,97 @@ func (s *Scorch) RollbackPoints() ([]*RollbackPoint, error) {
 	return rollbackPoints, nil
 }
 
-// Rollback atomically and durably (if unsafeBatch is unset) brings
-// the store back to the point in time as represented by the
-// RollbackPoint. Rollback() should only be passed a RollbackPoint
-// that came from the same store using the RollbackPoints() API.
-func (s *Scorch) Rollback(to *RollbackPoint) error {
+// Rollback atomically and durably brings the store back to the point
+// in time as represented by the RollbackPoint.
+// Rollback() should only be passed a RollbackPoint that came from the
+// same store using the RollbackPoints() API along with the index path.
+func Rollback(path string, to *RollbackPoint) error {
 	if to == nil {
 		return fmt.Errorf("Rollback: RollbackPoint is nil")
 	}
-
-	if s.rootBolt == nil {
-		return fmt.Errorf("Rollback: root is nil")
+	if len(path) == 0 {
+		return fmt.Errorf("Rollback: index path is empty")
 	}
 
-	revert := &snapshotReversion{}
-
-	s.rootLock.Lock()
+	rootBoltPath := path + string(os.PathSeparator) + "root.bolt"
+	rootBoltOpt := &bolt.Options{
+		ReadOnly: false,
+	}
+	rootBolt, err := bolt.Open(rootBoltPath, 0600, rootBoltOpt)
+	if err != nil || rootBolt == nil {
+		return err
+	}
+	defer func() {
+		err1 := rootBolt.Close()
+		if err1 != nil && err == nil {
+			err = err1
+		}
+	}()
 
-	err := s.rootBolt.View(func(tx *bolt.Tx) error {
+	// pick all the persisted epochs in bolt store
+	var found bool
+	var persistedEpochs []uint64
+	err = rootBolt.View(func(tx *bolt.Tx) error {
 		snapshots := tx.Bucket(boltSnapshotsBucket)
 		if snapshots == nil {
-			return fmt.Errorf("Rollback: no snapshots available")
-		}
-
-		pos := segment.EncodeUvarintAscending(nil, to.epoch)
-
-		snapshot := snapshots.Bucket(pos)
-		if snapshot == nil {
-			return fmt.Errorf("Rollback: snapshot not found")
+			return nil
 		}
-
-		indexSnapshot, err := s.loadSnapshot(snapshot)
-		if err != nil {
-			return fmt.Errorf("Rollback: unable to load snapshot: %v", err)
-		}
-
-		// add segments referenced by loaded index snapshot to the
-		// ineligibleForRemoval map
-		for _, segSnap := range indexSnapshot.segment {
-			filename := zapFileName(segSnap.id)
-			s.ineligibleForRemoval[filename] = true
+		sc := snapshots.Cursor()
+		for sk, _ := sc.Last(); sk != nil; sk, _ = sc.Prev() {
+			_, snapshotEpoch, err := segment.DecodeUvarintAscending(sk)
+			if err != nil {
+				continue
+			}
+			if snapshotEpoch == to.epoch {
+				found = true
```

Comment:
two thoughts here...
- if it's found, wonder if you can break out of this loop early?
- that is, instead of grabbing all the persistedEpochs
  wondering if you instead need the "epochs we're thinking that we're gonna delete"? 
- wondering if the == epoch check needs to be a >= or > kind of comparison instead?

Reply:
also, not sure if if bolt snapshot cursor needs to be closed?

Reply:
cursors doesn't seem to have a close method.

Reply:
> wondering if the == epoch check needs to be a >= or > kind of comparison instead?

I think, once we have the early exit logic that should be taken care.

Reply:
one thing about numbers is I tend to favor comparisons rather than equality...

like with equality, it ought to be ok when the epochs #'s are cleanly increasing by 1... like, 2,3,4,5,6,7, etc

but, if they're not only just +1 increasing (not sure why that might be the case) but if for whatever reason the bolt stuff has some missing #'s... and the epoch we're trying to get to is that #570 then the range comparison feels better to me?

Reply:
@steveyen on this one I disagree.  If you use a comparison operator here you are presuming something which may not be the case.  To me, the "fuzziness" part of, I want something in this range, is done externally when the caller evaluates which rollback points are available, and what metadata is associated with each point.  In this method call, they are choosing a specific rollback point of the known rollback points.  I think it should either rollback to exactly that point, or return an error if it cannot.

Reply:
oh, right!  headslap!  yah -- the evaluation is done by the caller of RollbackPoints.